Kudu.Core/SiteExtensions/SiteExtensionStatus.cs

```diff
@@ -104,6 +117,72 @@ public bool IsTerminalStatus()
                 || string.Equals(Constants.SiteExtensionProvisioningStateCanceled, ProvisioningState, StringComparison.OrdinalIgnoreCase);
         }
 
+        /// <summary>
+        /// <para>Loop thru all site extension settings files under 'siteExtensionStatusRoot', check if there is any successful installation</para>
+        /// <para>if not install to webroot, will trigger restart; if install to webroot and with applicationHost.xdt file, trigger restart.</para>
+        /// </summary>
+        /// <param name="siteExtensionStatusRoot">should be $ROOT\site\siteextensions</param>
+        /// <param name="siteExtensionRoot">should be $ROOT\SiteExtensions</param>
+        public static bool IsAnyInstallationRequireRestart(string siteExtensionStatusRoot, string siteExtensionRoot, ITracer tracer)
+        {
+            try
+            {
+                string[] packageDirs = FileSystemHelpers.GetDirectories(siteExtensionStatusRoot);
+                foreach (var dir in packageDirs)
+                {
+                    try
+                    {
+                        DirectoryInfo dirInfo = new DirectoryInfo(dir);
+                        string[] settingFile = FileSystemHelpers.GetFiles(dir, _statusSettingsFileName);
+                        foreach (var file in settingFile)
+                        {
+                            var statusSettings = new SiteExtensionStatus(siteExtensionStatusRoot, dirInfo.Name, tracer);
+                            if (IsSiteExtensionRequireRestart(statusSettings, siteExtensionRoot))
+                            {
+                                return true;
+                            }
+                        }
+                    }
+                    catch(Exception ex)
+                    {
+                        tracer.TraceError(ex);
+                        tracer.TraceError("Failed to query {0} under {1}, continus to check others ...", _statusSettingsFileName, dir);
```

Comment:
This could combine into one tracer.TraceError.   Please also make note to support writing to MDS (#1507) as it is unexpected exception.